api/src/main/java/edu/cornell/mannlib/vitro/webapp/dynapi/components/Parameter.java

```diff
@@ -52,8 +56,39 @@ public void addValidator(Validator validator) {
         validators.add(validator);
     }
 
-    public boolean isValid(String name, String[] values) {
-        return validators.isAllValid(name, values);
+    public boolean isValid(Data data) {
+        return isValid(name, data, type);
+    }
+
+    private boolean isValid(String name, Data data, ParameterType parameterType) {
+        boolean retVal = data != null && data.checkType(parameterType);
+        if (retVal) {
+            if (parameterType instanceof PrimitiveParameterType) {
```

Comment:
I would suggest to avoid conditionals with instancesOf checks by using more polymorphism where it makes sense.

Reply:
I agree, but don't know how to improve the code here. On one side there are Parameters and ParameterTypes defining the structure of data, on the another there are ObjectData, ArrayData, PrimitiveData, representing the data itself. I don't know how to implement some features without using instanceof.

Reply:
@litvinovg I found a way to replace some conditionals with polymorphism. Please check!